runtime/compiler/control/CompilationThread.hpp

```diff
@@ -204,35 +204,57 @@ class CompilationInfoPerThreadBase
    bool methodCanBeCompiled(TR_Memory *trMemory, TR_FrontEnd *fe, TR_ResolvedMethod *compilee, TR_FilterBST *&filter);
    int32_t                getCompThreadId() const { return _compThreadId; }
 
-   bool                   compilationCanBeInterrupted() const { return _compilationCanBeInterrupted; }
-   void                   enterInterruptibleOperation()
-      {
-      TR_ASSERT(!_compilationCanBeInterrupted, "This guard is not re-entrant.");
-      _compilationCanBeInterrupted = true;
-      }
-   void                   exitInterruptibleOperation()
-      {
-      TR_ASSERT(_compilationCanBeInterrupted, "Exiting interruptible operation without having entered.");
-      _compilationCanBeInterrupted = false;
-      }
+   bool compilationCanBeInterrupted() const { return _compilationCanBeInterrupted; }
 
    class InterruptibleOperation
       {
+      friend class TR::CompilationInfoPerThreadBase;
+      
    public:
       InterruptibleOperation(TR::CompilationInfoPerThreadBase &compThread) :
          _compThread(compThread)
          {
-         _compThread.enterInterruptibleOperation();
+         TR_ASSERT_FATAL(!_compThread._compilationCanBeInterrupted, "This guard is not re-entrant");
```

Comment:
Good point. The current implementation is incomplete. A boolean however will not suffice because we can have:

```
{
// Location 1
UninterruptibleOperation
   {
   // Location 2
   UninterruptibleOperation 
   }

...
   
   {
   // Location 3
   InterruptibleOperation 
   }
}
```

In which case if we use a boolean, there is no way to tell if there are any "outer" uninterruptible operations on the stack at `// Location 2`. We must use a depth counter instead.

Fixed in fa2769b.

Reply:
This might be nitpicky, but shouldn't we assert here that `_compThread._uninterruptableOperationDepth == 0` rather than assert in the destructor that `_compThread._compilationCanBeInterrupted == true`; nothing's wrong with asserting in the destructor but it would be more useful to assert sooner rather than later.

Reply:
> Nesting an InterruptibleOperation under an UninterruptibleOperation acts as a NOP

Except wouldn't we assert right now in the destructor? The constructor of `InterruptibleOperation` only sets ` _compThread._compilationCanBeInterrupted = true;` if `_compThread._uninterruptableOperationDepth == 0` which is not going to be the case if you nest an `InterruptibleOperation` under an `UninterruptibleOperation`. This means that, in `Location 3` when we call the destructor of `InterruptibleOperation` it will assert because `_compThread._compilationCanBeInterrupted` is not going to be true.

Reply:
> What do you think about removing the asserts and adding the extra documentation?

Yeah, that is probably the best way to go about this. Part of me worries about the possibility of some code somewhere hinging on the fact that it is within an interruptible scope, but I can't really think of why that would be necessary given that we always have that check before we interrupt.

Reply:
Fixed in b58db9d.